protocol/x/clob/keeper/deleveraging.go

```diff
@@ -499,23 +532,88 @@ func (k Keeper) ProcessDeleveraging(
 		),
 	)
 
-	// Send on-chain update for the deleveraging. The events are stored in a TransientStore which should be rolled-back
-	// if the branched state is discarded, so batching is not necessary.
-	k.GetIndexerEventManager().AddTxnEvent(
-		ctx,
-		indexerevents.SubtypeDeleveraging,
-		indexerevents.DeleveragingEventVersion,
-		indexer_manager.GetBytes(
-			indexerevents.NewDeleveragingEvent(
-				liquidatedSubaccountId,
-				offsettingSubaccountId,
-				perpetualId,
-				satypes.BaseQuantums(new(big.Int).Abs(deltaBaseQuantums).Uint64()),
-				satypes.BaseQuantums(deltaQuoteQuantums.Uint64()),
-				deltaBaseQuantums.Sign() > 0,
-			),
-		),
+	return nil
+}
+
+// GetSubaccountsWithPositionsInFinalSettlementMarkets uses the subaccountOpenPositionInfo returned from the
+// liquidations daemon to fetch subaccounts with open positions in final settlement markets. These subaccounts
+// will be deleveraged in either at the oracle price if non-negative TNC or at bankruptcy price if negative TNC. This
+// function is called in PrepareCheckState during the deleveraging step.
+func (k Keeper) GetSubaccountsWithPositionsInFinalSettlementMarkets(
+	ctx sdk.Context,
+	subaccountOpenPositionInfo map[uint32]*types.SubaccountOpenPositionInfo,
+) (subaccountsToDeleverage []subaccountToDeleverage) {
+	defer telemetry.MeasureSince(
+		time.Now(),
+		types.ModuleName,
+		metrics.ClobGetSubaccountsWithPositionsInFinalSettlementMarkets,
+		metrics.Latency,
+	)
+
+	for _, clobPair := range k.GetAllClobPairs(ctx) {
+		if clobPair.Status != types.ClobPair_STATUS_FINAL_SETTLEMENT {
+			continue
+		}
+
+		finalSettlementPerpetualId := clobPair.MustGetPerpetualId()
+		positionInfo, found := subaccountOpenPositionInfo[finalSettlementPerpetualId]
+		if !found {
+			// No open positions in the market.
+			continue
+		}
+
+		for _, subaccountId := range positionInfo.SubaccountsWithLongPosition {
+			subaccountsToDeleverage = append(subaccountsToDeleverage, subaccountToDeleverage{
+				SubaccountId: subaccountId,
+				PerpetualId:  finalSettlementPerpetualId,
+			})
+		}
+		for _, subaccountId := range positionInfo.SubaccountsWithShortPosition {
+			subaccountsToDeleverage = append(subaccountsToDeleverage, subaccountToDeleverage{
+				SubaccountId: subaccountId,
+				PerpetualId:  finalSettlementPerpetualId,
+			})
+		}
+	}
+
+	metrics.AddSample(
+		metrics.ClobSubaccountsWithFinalSettlementPositionsCount,
+		float32(len(subaccountsToDeleverage)),
 	)
 
+	return subaccountsToDeleverage
+}
+
+// DeleverageSubaccounts deleverages a slice of subaccounts paired with a perpetual position to deleverage with.
+// Returns an error if a deleveraging attempt returns an error.
+func (k Keeper) DeleverageSubaccounts(
+	ctx sdk.Context,
+	subaccountsToDeleverage []subaccountToDeleverage,
+) error {
+	defer func() {
+		telemetry.MeasureSince(
+			time.Now(),
+			types.ModuleName,
+			metrics.LiquidateSubaccounts_Deleverage,
+			metrics.Latency,
+		)
+	}()
```

Comment:
I don't think this works. time.Now() will get evaluated when defer function runs

Reply:
Why can't this just be the following (so func args are evaluated right away)?

```suggestion
	defer telemetry.MeasureSince(
			time.Now(),
			types.ModuleName,
			metrics.LiquidateSubaccounts_Deleverage,
			metrics.Latency,
		)
	}
```

Reply:
yeah that's my bad, Lucas' suggestion is right